{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare feature engineering checklists for the 6 feature categories:\n",
    "- 01 Functional Annotations  \n",
    "- 02 Evolutionary Properties  \n",
    "- 03 Protein sequence properties  \n",
    "- 04 Gene expression  \n",
    "- 05 Epigenetics  \n",
    "- 06 Network properties  \n",
    "The feature data is coming from the paper by Cusack et al., 2021 MBE and is located \n",
    "at /home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature category</th>\n",
       "      <th>Feature name</th>\n",
       "      <th>NEW Feature name</th>\n",
       "      <th>ML model feature name</th>\n",
       "      <th>Feature type</th>\n",
       "      <th>Calculation for gene pair</th>\n",
       "      <th>Data processing method</th>\n",
       "      <th>Transformation</th>\n",
       "      <th>Description</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Epigenetic modifications</td>\n",
       "      <td>Gene body methylation</td>\n",
       "      <td>gene body methylation</td>\n",
       "      <td>continuous_bodyMethylation.MLD_number_in_pair_log</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>A gene was called as gene body methylated if t...</td>\n",
       "      <td>Lloyd et al. 2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Epigenetic modifications</td>\n",
       "      <td>Gene body methylation</td>\n",
       "      <td>gene body methylation</td>\n",
       "      <td>continuous_bodyMethylation.MLD_number_in_pair_...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>A gene was called as gene body methylated if t...</td>\n",
       "      <td>Lloyd et al. 2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Epigenetic modifications</td>\n",
       "      <td>Gene body methylation</td>\n",
       "      <td>gene body methylation</td>\n",
       "      <td>continuous_bodyMethylation.MLD_number_in_pair_...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Reciprocal</td>\n",
       "      <td>A gene was called as gene body methylated if t...</td>\n",
       "      <td>Lloyd et al. 2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Epigenetic modifications</td>\n",
       "      <td>Gene body methylation</td>\n",
       "      <td>gene body methylation</td>\n",
       "      <td>continuous_bodyMethylation.MLD_number_in_pair_...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Squared</td>\n",
       "      <td>A gene was called as gene body methylated if t...</td>\n",
       "      <td>Lloyd et al. 2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Epigenetic modifications</td>\n",
       "      <td>Chromatin accessibility</td>\n",
       "      <td>chromatin accessibility</td>\n",
       "      <td>continuous_chromatin_accessibility_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>Whether or not gene contained a Dnase peak site</td>\n",
       "      <td>Sullivan et al. 2014</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4111</th>\n",
       "      <td>Protein properties</td>\n",
       "      <td>Protein domain annotations</td>\n",
       "      <td>protein domain annotations</td>\n",
       "      <td>continuous_protein_domain_total_binned</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Total #</td>\n",
       "      <td>Categorical</td>\n",
       "      <td>Binned</td>\n",
       "      <td>All protein domain annotations for a gene.</td>\n",
       "      <td>Finn et al. 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4112</th>\n",
       "      <td>Protein properties</td>\n",
       "      <td>Protein domain annotations</td>\n",
       "      <td>protein domain annotations</td>\n",
       "      <td>continuous_protein_domain_total_log</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Total #</td>\n",
       "      <td>Categorical</td>\n",
       "      <td>Log</td>\n",
       "      <td>All protein domain annotations for a gene.</td>\n",
       "      <td>Finn et al. 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4113</th>\n",
       "      <td>Protein properties</td>\n",
       "      <td>Protein domain annotations</td>\n",
       "      <td>protein domain annotations</td>\n",
       "      <td>continuous_protein_domain_total_noTF</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Total #</td>\n",
       "      <td>Categorical</td>\n",
       "      <td>NaN</td>\n",
       "      <td>All protein domain annotations for a gene.</td>\n",
       "      <td>Finn et al. 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4114</th>\n",
       "      <td>Protein properties</td>\n",
       "      <td>Protein domain annotations</td>\n",
       "      <td>protein domain annotations</td>\n",
       "      <td>continuous_protein_domain_total_reciprocal</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Total #</td>\n",
       "      <td>Categorical</td>\n",
       "      <td>Reciprocal</td>\n",
       "      <td>All protein domain annotations for a gene.</td>\n",
       "      <td>Finn et al. 2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4115</th>\n",
       "      <td>Protein properties</td>\n",
       "      <td>Protein domain annotations</td>\n",
       "      <td>protein domain annotations</td>\n",
       "      <td>continuous_protein_domain_total_squared</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Total #</td>\n",
       "      <td>Categorical</td>\n",
       "      <td>Squared</td>\n",
       "      <td>All protein domain annotations for a gene.</td>\n",
       "      <td>Finn et al. 2016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4116 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              Feature category                Feature name  \\\n",
       "0     Epigenetic modifications       Gene body methylation   \n",
       "1     Epigenetic modifications       Gene body methylation   \n",
       "2     Epigenetic modifications       Gene body methylation   \n",
       "3     Epigenetic modifications       Gene body methylation   \n",
       "4     Epigenetic modifications     Chromatin accessibility   \n",
       "...                        ...                         ...   \n",
       "4111        Protein properties  Protein domain annotations   \n",
       "4112        Protein properties  Protein domain annotations   \n",
       "4113        Protein properties  Protein domain annotations   \n",
       "4114        Protein properties  Protein domain annotations   \n",
       "4115        Protein properties  Protein domain annotations   \n",
       "\n",
       "                NEW Feature name  \\\n",
       "0          gene body methylation   \n",
       "1          gene body methylation   \n",
       "2          gene body methylation   \n",
       "3          gene body methylation   \n",
       "4        chromatin accessibility   \n",
       "...                          ...   \n",
       "4111  protein domain annotations   \n",
       "4112  protein domain annotations   \n",
       "4113  protein domain annotations   \n",
       "4114  protein domain annotations   \n",
       "4115  protein domain annotations   \n",
       "\n",
       "                                  ML model feature name Feature type  \\\n",
       "0     continuous_bodyMethylation.MLD_number_in_pair_log   Continuous   \n",
       "1     continuous_bodyMethylation.MLD_number_in_pair_...   Continuous   \n",
       "2     continuous_bodyMethylation.MLD_number_in_pair_...   Continuous   \n",
       "3     continuous_bodyMethylation.MLD_number_in_pair_...   Continuous   \n",
       "4     continuous_chromatin_accessibility_number_in_p...   Continuous   \n",
       "...                                                 ...          ...   \n",
       "4111             continuous_protein_domain_total_binned   Continuous   \n",
       "4112                continuous_protein_domain_total_log   Continuous   \n",
       "4113               continuous_protein_domain_total_noTF   Continuous   \n",
       "4114         continuous_protein_domain_total_reciprocal   Continuous   \n",
       "4115            continuous_protein_domain_total_squared   Continuous   \n",
       "\n",
       "     Calculation for gene pair Data processing method Transformation  \\\n",
       "0               Number in pair                 Binary            Log   \n",
       "1               Number in pair                 Binary            NaN   \n",
       "2               Number in pair                 Binary     Reciprocal   \n",
       "3               Number in pair                 Binary        Squared   \n",
       "4               Number in pair                 Binary            Log   \n",
       "...                        ...                    ...            ...   \n",
       "4111                   Total #            Categorical         Binned   \n",
       "4112                   Total #            Categorical            Log   \n",
       "4113                   Total #            Categorical            NaN   \n",
       "4114                   Total #            Categorical     Reciprocal   \n",
       "4115                   Total #            Categorical        Squared   \n",
       "\n",
       "                                            Description                Source  \n",
       "0     A gene was called as gene body methylated if t...     Lloyd et al. 2015  \n",
       "1     A gene was called as gene body methylated if t...     Lloyd et al. 2015  \n",
       "2     A gene was called as gene body methylated if t...     Lloyd et al. 2015  \n",
       "3     A gene was called as gene body methylated if t...     Lloyd et al. 2015  \n",
       "4       Whether or not gene contained a Dnase peak site  Sullivan et al. 2014  \n",
       "...                                                 ...                   ...  \n",
       "4111         All protein domain annotations for a gene.      Finn et al. 2016  \n",
       "4112         All protein domain annotations for a gene.      Finn et al. 2016  \n",
       "4113         All protein domain annotations for a gene.      Finn et al. 2016  \n",
       "4114         All protein domain annotations for a gene.      Finn et al. 2016  \n",
       "4115         All protein domain annotations for a gene.      Finn et al. 2016  \n",
       "\n",
       "[4116 rows x 10 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cusack's feature list\n",
    "master_list = pd.read_excel('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_tables_revision.xlsx',\n",
    "                            sheet_name='Supplemental table 1', skipfooter=22)\n",
    "master_list = master_list[['Feature category', 'Feature name', 'ML model feature name',\n",
    "             'Feature type', 'Calculation for gene pair', 'Data processing method',\n",
    "             'Transformation', 'Description', 'Source']]\n",
    "master_list.insert(2, 'NEW Feature name', master_list['Feature name'].str.strip().str.lower()) # Add new column with lowercase feature name\n",
    "# master_list.to_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_table_1.csv', index=False)\n",
    "master_list"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in the master list with corrected 'Feature name' values\n",
    "July 29 - Functional Annotation feature names were corrected, see `01_functional_annotation features` section  \n",
    "July 30 - "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 01_functional_annotation features (July 29)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_list = pd.read_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_table_1_corrected_KSA_20240729.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>File path</th>\n",
       "      <th>Feature name</th>\n",
       "      <th>Feature category</th>\n",
       "      <th>NEW Feature name</th>\n",
       "      <th>ML model feature name</th>\n",
       "      <th>Feature type</th>\n",
       "      <th>Calculation for gene pair</th>\n",
       "      <th>Data processing method</th>\n",
       "      <th>Transformation</th>\n",
       "      <th>Description</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Reciprocal</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Squared</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,4-dihydroxy-2-naphthoate biosynthesis ii plants</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,4-dihydroxy-2-naphthoate biosynthesis ii plants</td>\n",
       "      <td>continuous_1.4.dihydroxy.2.naphthoate_biosynth...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin, antheraxanthin and violaxanthin in...</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin, antheraxanthin and violaxanthin in...</td>\n",
       "      <td>continuous_zeaxanthin._antheraxanthin_and_viol...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Squared</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2426</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2427</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2428</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Reciprocal</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2429</th>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Squared</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2430 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              File path  \\\n",
       "0     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "1     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "3     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "4     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "...                                                 ...   \n",
       "2425  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2426  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2427  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2428  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2429  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "\n",
       "                                           Feature name  \\\n",
       "0                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "1                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "2                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "3                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "4     1,4-dihydroxy-2-naphthoate biosynthesis ii plants   \n",
       "...                                                 ...   \n",
       "2425  zeaxanthin, antheraxanthin and violaxanthin in...   \n",
       "2426                            zeaxanthin biosynthesis   \n",
       "2427                            zeaxanthin biosynthesis   \n",
       "2428                            zeaxanthin biosynthesis   \n",
       "2429                            zeaxanthin biosynthesis   \n",
       "\n",
       "           Feature category  \\\n",
       "0     Functional annotation   \n",
       "1     Functional annotation   \n",
       "2     Functional annotation   \n",
       "3     Functional annotation   \n",
       "4     Functional annotation   \n",
       "...                     ...   \n",
       "2425  Functional annotation   \n",
       "2426  Functional annotation   \n",
       "2427  Functional annotation   \n",
       "2428  Functional annotation   \n",
       "2429  Functional annotation   \n",
       "\n",
       "                                       NEW Feature name  \\\n",
       "0                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "1                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "2                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "3                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "4     1,4-dihydroxy-2-naphthoate biosynthesis ii plants   \n",
       "...                                                 ...   \n",
       "2425  zeaxanthin, antheraxanthin and violaxanthin in...   \n",
       "2426                            zeaxanthin biosynthesis   \n",
       "2427                            zeaxanthin biosynthesis   \n",
       "2428                            zeaxanthin biosynthesis   \n",
       "2429                            zeaxanthin biosynthesis   \n",
       "\n",
       "                                  ML model feature name Feature type  \\\n",
       "0     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "1     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "2     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "3     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "4     continuous_1.4.dihydroxy.2.naphthoate_biosynth...   Continuous   \n",
       "...                                                 ...          ...   \n",
       "2425  continuous_zeaxanthin._antheraxanthin_and_viol...   Continuous   \n",
       "2426  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "2427  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "2428  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "2429  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "\n",
       "     Calculation for gene pair Data processing method Transformation  \\\n",
       "0               Number in pair                 Binary            Log   \n",
       "1               Number in pair                 Binary            NaN   \n",
       "2               Number in pair                 Binary     Reciprocal   \n",
       "3               Number in pair                 Binary        Squared   \n",
       "4               Number in pair                 Binary            Log   \n",
       "...                        ...                    ...            ...   \n",
       "2425            Number in pair                 Binary        Squared   \n",
       "2426            Number in pair                 Binary            Log   \n",
       "2427            Number in pair                 Binary            NaN   \n",
       "2428            Number in pair                 Binary     Reciprocal   \n",
       "2429            Number in pair                 Binary        Squared   \n",
       "\n",
       "                                            Description  \\\n",
       "0     Gene product involved in this pathway or funct...   \n",
       "1     Gene product involved in this pathway or funct...   \n",
       "2     Gene product involved in this pathway or funct...   \n",
       "3     Gene product involved in this pathway or funct...   \n",
       "4     Gene product involved in this pathway or funct...   \n",
       "...                                                 ...   \n",
       "2425  Gene product involved in this pathway or funct...   \n",
       "2426  Gene product involved in this pathway or funct...   \n",
       "2427  Gene product involved in this pathway or funct...   \n",
       "2428  Gene product involved in this pathway or funct...   \n",
       "2429  Gene product involved in this pathway or funct...   \n",
       "\n",
       "                                Source  \n",
       "0     AraCyc v.15; Mueller et al. 2003  \n",
       "1     AraCyc v.15; Mueller et al. 2003  \n",
       "2     AraCyc v.15; Mueller et al. 2003  \n",
       "3     AraCyc v.15; Mueller et al. 2003  \n",
       "4     AraCyc v.15; Mueller et al. 2003  \n",
       "...                                ...  \n",
       "2425  AraCyc v.15; Mueller et al. 2003  \n",
       "2426  AraCyc v.15; Mueller et al. 2003  \n",
       "2427  AraCyc v.15; Mueller et al. 2003  \n",
       "2428  AraCyc v.15; Mueller et al. 2003  \n",
       "2429  AraCyc v.15; Mueller et al. 2003  \n",
       "\n",
       "[2430 rows x 11 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get all the JSON file paths\n",
    "files = []\n",
    "path = '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/01_functional_annotation/AraCyc_pathways/AraCyc'\n",
    "for dirpath, dirnames, filenames in os.walk(path):\n",
    "    for filename in filenames:\n",
    "        files.append(os.path.join(dirpath, filename))\n",
    "len(files) # 608\n",
    "\n",
    "# Create a dataframe with the file paths and feature names\n",
    "file_paths = pd.DataFrame(np.repeat(files, 4)).sort_values(0) # repeat each file path 4 times since there are 4 data transformations per feature\n",
    "feat_names = file_paths.iloc[:,0].str.split('/').apply(lambda x: x[10]).\\\n",
    "    apply(lambda x: x.split('_dictionary.json')[0].replace('_', ' ')) # get the feature name from the JSON file name\n",
    "file_paths.insert(1, 'Feature name', feat_names) # insert the feature name into the dataframe\n",
    "file_paths.columns = ['File path', 'Feature name']\n",
    "file_paths['Feature name'] = file_paths['Feature name'].str.lower()\n",
    "\n",
    "# Map the file paths to the master list of features\n",
    "func_anno_paths = file_paths.merge(master_list.loc[master_list['Feature category']\\\n",
    "    =='Functional annotation'], left_on='Feature name', right_on='NEW Feature name',\n",
    "    how='left', suffixes=('', '_y')) # merge to master list\n",
    "func_anno_paths.drop(columns=['Feature name_y'], inplace=True)\n",
    "\n",
    "# Drop duplicates created by merge\n",
    "func_anno_paths = func_anno_paths.drop_duplicates('ML model feature name').reset_index().drop('index', axis=1)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "File path                      0\n",
       "Feature name                   0\n",
       "Feature category               1\n",
       "NEW Feature name               1\n",
       "ML model feature name          1\n",
       "Feature type                   1\n",
       "Calculation for gene pair      1\n",
       "Data processing method         1\n",
       "Transformation               608\n",
       "Description                    1\n",
       "Source                         1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check for missing data\n",
    "func_anno_paths.isna().sum()\n",
    "# Missing data in original master list file from Cusack 2021 paper\n",
    "# File path                       0\n",
    "# Feature name                    0\n",
    "# Feature category               96\n",
    "# ML model feature name          96\n",
    "# Feature type                   96\n",
    "# Calculation for gene pair      96\n",
    "# Data processing method         96\n",
    "# Transformation               2436\n",
    "# Description                    96\n",
    "# Source                         96\n",
    "# dtype: int64"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1d-myo-inositol hexakisphosphate biosynthesis v from ins1,3,4p3'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Manually check that these features were not included in the master listfunc_anno_paths.loc[func_anno_paths['ML model feature name'].isna(),'Feature name'].unique()\n",
    "func_anno_paths.loc[func_anno_paths['ML model feature name'].isna(),'Feature name'].unique()\n",
    "\n",
    "# Note: There are 96 rows with missing data in the original master list file from Cusack 2021 paper\n",
    "# Some of these have (), sub2sub, sub3sub, sub12sub, etc. in the master list. I \n",
    "# modified the file Supplemental_table_1_corrected_KSA_20240729.csv to correct \n",
    "# the errors. See 'NEW Feature name' column in the master list file for corrected \n",
    "# feature names.\n",
    "#\n",
    "# array(['1d-myo-inositol hexakisphosphate biosynthesis v from ins1,3,4p3', # not found in master list, thus will not be included as a feature\n",
    "#        '3e-4,8-dimethylnona-1,3,7-triene biosynthesis',                   # modified '(3e)' to '3e' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        '4-methyl-5beta-hydroxyethylthiazole salvage yeast',               # modified '5beta' to '5-beta' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'cosub2sub fixation into oxaloacetate anapleurotic',               # modified 'co2 fixation into oxaloacetate anapleurotic' to 'cosub2sub fixation into oxaloacetate anapleurotic' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'e,e-4,8,12-trimethyltrideca-1,3,7,11-tetraene biosynthesis',      # modified '(e,e)' to 'e,e' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'gasub12sub biosynthesis',                                         # modified 'ga12' to 'gasub12sub' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'kdo transfer to lipid ivsubasub i',                               # modified 'iva' to 'ivsubasub' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'l-nsupdeltasup-acetylornithine biosynthesis',                     # modified 'ndelta' to 'nsupdeltasup' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'nadnadh phosphorylation and dephosphorylation',                   # modified 'nad/nadh' to 'nadnadh' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'fatty acid beta-oxidation v unsaturated, odd number, di-isomerase-dependent', # modified '(unsaturated, odd number)' to 'unsaturated, odd number'\n",
    "#        'formate oxidation to cosub2sub',                                  # modified 'co2' to 'cosub2sub' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'gibberellin biosynthesis iii early c-13 hydroxylation',           # modified '(early c-13 hydroxylation)' to 'early c-13 hydroxylation' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'gibberellin biosynthesis ii early c-3 hydroxylation',             # modified '(early c-3 hydroxylation)' to 'early c-3 hydroxylation' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'gibberellin biosynthesis i non c-3, non c-13 hydroxylation',      # modified '(non c-3, non c-13 hydroxylation)' to 'non c-3, non c-13 hydroxylation' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'gibberellin inactivation i 2beta-hydroxylation',                  # modified '(2beta-hydroxylation)' to '2beta-hydroxylation' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'lipid-dependent phytate biosynthesis ii via ins1,3,4psub3sub',    # modified 'ins(1,3,4)p3' to 'ins1,3,4psub3sub' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'lipid-dependent phytate biosynthesis i via ins1,4,5psub3sub',     # modified 'ins(1,4,5)p3' to 'ins1,4,5psub3sub' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'lipid ivsubasub biosynthesis',                                    # modified 'iva' to 'ivsubasub' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'pyruvate decarboxylation to acetyl coa',                          # modified 'acetyl-coa' to 'acetyl coa' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'superpathay of heme biosynthesis from glutamate',                 # modified 'superpathway' to 'superpathay' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'superpathway of gibberellin gasub12sub biosynthesis',             # modified 'ga12' to 'gasub12sub' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'superpathway of pantothenate and coenzymea biosynthesis',         # modified 'coenzyme a' to 'coenzymea' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'trans, trans-farnesyl diphosphate biosynthesis',                  # modified 'trans-trans' to 'trans, trans' in Supplemental_table_1_corrected_KSA_20240729.csv\n",
    "#        'traumatin and z-3-hexen-1-yl acetate biosynthesis'], dtype=object)# modified '(z)' to 'z' in Supplemental_table_1_corrected_KSA_20240729.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Created by</th>\n",
       "      <th>Date created</th>\n",
       "      <th>File path</th>\n",
       "      <th>Feature name</th>\n",
       "      <th>Feature category</th>\n",
       "      <th>NEW Feature name</th>\n",
       "      <th>ML model feature name</th>\n",
       "      <th>Feature type</th>\n",
       "      <th>Calculation for gene pair</th>\n",
       "      <th>Data processing method</th>\n",
       "      <th>Transformation</th>\n",
       "      <th>Description</th>\n",
       "      <th>Source</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Reciprocal</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,3,5-trimethoxybenzene biosynthesis</td>\n",
       "      <td>continuous_1.3.5.trimethoxybenzene_biosynthesi...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Squared</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>1,4-dihydroxy-2-naphthoate biosynthesis ii plants</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>1,4-dihydroxy-2-naphthoate biosynthesis ii plants</td>\n",
       "      <td>continuous_1.4.dihydroxy.2.naphthoate_biosynth...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2425</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin, antheraxanthin and violaxanthin in...</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin, antheraxanthin and violaxanthin in...</td>\n",
       "      <td>continuous_zeaxanthin._antheraxanthin_and_viol...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Squared</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2426</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Log</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2427</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>None</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2428</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Reciprocal</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2429</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>/home/seguraab/ara-kinase-prediction/data/2021...</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>Functional annotation</td>\n",
       "      <td>zeaxanthin biosynthesis</td>\n",
       "      <td>continuous_zeaxanthin_biosynthesis_number_in_p...</td>\n",
       "      <td>Continuous</td>\n",
       "      <td>Number in pair</td>\n",
       "      <td>Binary</td>\n",
       "      <td>Squared</td>\n",
       "      <td>Gene product involved in this pathway or funct...</td>\n",
       "      <td>AraCyc v.15; Mueller et al. 2003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2430 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Created by Date created  \\\n",
       "0                              \n",
       "1                              \n",
       "2                              \n",
       "3                              \n",
       "4                              \n",
       "...         ...          ...   \n",
       "2425                           \n",
       "2426                           \n",
       "2427                           \n",
       "2428                           \n",
       "2429                           \n",
       "\n",
       "                                              File path  \\\n",
       "0     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "1     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "3     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "4     /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "...                                                 ...   \n",
       "2425  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2426  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2427  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2428  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "2429  /home/seguraab/ara-kinase-prediction/data/2021...   \n",
       "\n",
       "                                           Feature name  \\\n",
       "0                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "1                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "2                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "3                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "4     1,4-dihydroxy-2-naphthoate biosynthesis ii plants   \n",
       "...                                                 ...   \n",
       "2425  zeaxanthin, antheraxanthin and violaxanthin in...   \n",
       "2426                            zeaxanthin biosynthesis   \n",
       "2427                            zeaxanthin biosynthesis   \n",
       "2428                            zeaxanthin biosynthesis   \n",
       "2429                            zeaxanthin biosynthesis   \n",
       "\n",
       "           Feature category  \\\n",
       "0     Functional annotation   \n",
       "1     Functional annotation   \n",
       "2     Functional annotation   \n",
       "3     Functional annotation   \n",
       "4     Functional annotation   \n",
       "...                     ...   \n",
       "2425  Functional annotation   \n",
       "2426  Functional annotation   \n",
       "2427  Functional annotation   \n",
       "2428  Functional annotation   \n",
       "2429  Functional annotation   \n",
       "\n",
       "                                       NEW Feature name  \\\n",
       "0                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "1                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "2                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "3                  1,3,5-trimethoxybenzene biosynthesis   \n",
       "4     1,4-dihydroxy-2-naphthoate biosynthesis ii plants   \n",
       "...                                                 ...   \n",
       "2425  zeaxanthin, antheraxanthin and violaxanthin in...   \n",
       "2426                            zeaxanthin biosynthesis   \n",
       "2427                            zeaxanthin biosynthesis   \n",
       "2428                            zeaxanthin biosynthesis   \n",
       "2429                            zeaxanthin biosynthesis   \n",
       "\n",
       "                                  ML model feature name Feature type  \\\n",
       "0     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "1     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "2     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "3     continuous_1.3.5.trimethoxybenzene_biosynthesi...   Continuous   \n",
       "4     continuous_1.4.dihydroxy.2.naphthoate_biosynth...   Continuous   \n",
       "...                                                 ...          ...   \n",
       "2425  continuous_zeaxanthin._antheraxanthin_and_viol...   Continuous   \n",
       "2426  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "2427  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "2428  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "2429  continuous_zeaxanthin_biosynthesis_number_in_p...   Continuous   \n",
       "\n",
       "     Calculation for gene pair Data processing method Transformation  \\\n",
       "0               Number in pair                 Binary            Log   \n",
       "1               Number in pair                 Binary           None   \n",
       "2               Number in pair                 Binary     Reciprocal   \n",
       "3               Number in pair                 Binary        Squared   \n",
       "4               Number in pair                 Binary            Log   \n",
       "...                        ...                    ...            ...   \n",
       "2425            Number in pair                 Binary        Squared   \n",
       "2426            Number in pair                 Binary            Log   \n",
       "2427            Number in pair                 Binary           None   \n",
       "2428            Number in pair                 Binary     Reciprocal   \n",
       "2429            Number in pair                 Binary        Squared   \n",
       "\n",
       "                                            Description  \\\n",
       "0     Gene product involved in this pathway or funct...   \n",
       "1     Gene product involved in this pathway or funct...   \n",
       "2     Gene product involved in this pathway or funct...   \n",
       "3     Gene product involved in this pathway or funct...   \n",
       "4     Gene product involved in this pathway or funct...   \n",
       "...                                                 ...   \n",
       "2425  Gene product involved in this pathway or funct...   \n",
       "2426  Gene product involved in this pathway or funct...   \n",
       "2427  Gene product involved in this pathway or funct...   \n",
       "2428  Gene product involved in this pathway or funct...   \n",
       "2429  Gene product involved in this pathway or funct...   \n",
       "\n",
       "                                Source  \n",
       "0     AraCyc v.15; Mueller et al. 2003  \n",
       "1     AraCyc v.15; Mueller et al. 2003  \n",
       "2     AraCyc v.15; Mueller et al. 2003  \n",
       "3     AraCyc v.15; Mueller et al. 2003  \n",
       "4     AraCyc v.15; Mueller et al. 2003  \n",
       "...                                ...  \n",
       "2425  AraCyc v.15; Mueller et al. 2003  \n",
       "2426  AraCyc v.15; Mueller et al. 2003  \n",
       "2427  AraCyc v.15; Mueller et al. 2003  \n",
       "2428  AraCyc v.15; Mueller et al. 2003  \n",
       "2429  AraCyc v.15; Mueller et al. 2003  \n",
       "\n",
       "[2430 rows x 13 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make the checklist file\n",
    "func_anno_paths['Transformation'] = func_anno_paths['Transformation'].fillna('None') # set NaNs in Transformation to \"None\"\n",
    "func_anno_paths.insert(0, 'Created by', '')\n",
    "func_anno_paths.insert(1, 'Date created', '')\n",
    "# func_anno_paths.to_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/01_functional_annotation_feature_list.csv', index=False)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 02_evolutionary_properties features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_list = pd.read_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_table_1_corrected_KSA_20240729.csv')\n",
    "\n",
    "# Subset the master list to get the protein properties\n",
    "evo_prop_df = master_list.loc[master_list['Feature category']=='Evolutionary properties',:]\n",
    "\n",
    "# JSON file paths for evolutionary properties\n",
    "# repeat these 25\n",
    "file_paths = ['/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/02_evolutionary_properties/gene_family_size/gene_family_size_dictionary.json',\n",
    "            #   'ka/ks', # ask Thilanka for a script to calculate these\n",
    "            #   'ka',\n",
    "            #   'ks',\n",
    "              '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/02_evolutionary_properties/lethality_score/lethality_dict_121817.json',\n",
    "              '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/02_evolutionary_properties/retention_rate/Retention_rate_dictionary.json']\n",
    "file_paths = file_paths + list(np.repeat('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/02_evolutionary_properties/lethality_score/lethality_dict_121817.json', 5))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 03_protein_sequence_properties features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_list = pd.read_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_table_1_corrected_KSA_20240729.csv')\n",
    "\n",
    "# Subset the master list to get the protein properties\n",
    "prot_prop_df = master_list.loc[master_list['Feature category']=='Protein properties',:]\n",
    "\n",
    "# JSON file paths for protein properties\n",
    "file_paths = list(np.repeat(['/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/AA_length/aaLength.continuous.MLD_dictionary.json',\n",
    "                             '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/isoelectric_point/isoelectric_point_dictionary.json'], 25))\n",
    "file_paths = file_paths + list(np.repeat(['/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/acetylation/acetylation_dict.json',\n",
    "                   '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/deamination/deamination_dict.json',\n",
    "                   '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/formylation/formylation_dict.json',\n",
    "                   '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/hydroxylation/hydroxylation_dict.json',\n",
    "                   '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/myristoylation/myristoylation_dict.json',\n",
    "                   '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/oxidation/oxidation_dict.json',\n",
    "                   '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/propionylation/propionylation_dict.json'], 4))\n",
    "file_paths = file_paths + list(np.repeat('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/protein_domain/pfam', 15))\n",
    "\n",
    "print(len(file_paths), prot_prop_df.shape) # 93 features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dataframe with the file paths and feature names  \n",
    "file_paths = pd.DataFrame(file_paths).sort_values(0)\n",
    "feat_names = file_paths.iloc[:,0].str.split('/').apply(lambda x: x[9]).\\\n",
    "    apply(lambda x: x.split('_dict.json')[0].replace('_', ' ')) # get the feature name from the JSON file name\n",
    "feat_names = feat_names.apply(lambda x: x.replace(' dictionary.json', ''))\n",
    "file_paths.insert(1, 'Feature name', feat_names) # insert the feature name into the dataframe\n",
    "file_paths.columns = ['File path', 'Feature name']\n",
    "file_paths['Feature name'] = file_paths['Feature name'].str.lower()\n",
    "\n",
    "# Map the file paths to the master list of features\n",
    "prot_prop_df.loc[prot_prop_df['Feature name']=='AA length', 'NEW Feature name'] = 'aalength.continuous.mld'\n",
    "prot_prop_df.loc[prot_prop_df['Feature name']=='Protein domain annotations', 'NEW Feature name'] = 'pfam'\n",
    "\n",
    "func_anno_paths = file_paths.merge(prot_prop_df, left_on='Feature name', right_on='NEW Feature name',\n",
    "    how='left', suffixes=('', '_y')) # merge to master list\n",
    "func_anno_paths.drop(columns=['Feature name'], inplace=True)\n",
    "\n",
    "# Drop duplicates created by merge\n",
    "func_anno_paths = func_anno_paths.drop_duplicates('ML model feature name').reset_index().drop('index', axis=1)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the checklist file\n",
    "func_anno_paths['Transformation'] = func_anno_paths['Transformation'].fillna('None') # set NaNs in Transformation to \"None\"\n",
    "func_anno_paths.insert(0, 'Created by', '')\n",
    "func_anno_paths.insert(1, 'Date created', '')\n",
    "func_anno_paths.rename(columns={'Feature name_y':'Feature name'}, inplace=True)\n",
    "# func_anno_paths.to_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/03_protein_sequence_properties_feature_list.csv', index=False)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 04_gene_expression features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_list = pd.read_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_table_1_corrected_KSA_20240729.csv')\n",
    "\n",
    "# Subset the master list to get the protein properties\n",
    "gene_expr_df = master_list.loc[master_list['Feature category']=='Gene expression',:]\n",
    "print(gene_expr_df.shape) # 450\n",
    "\n",
    "# Get all the JSON file paths\n",
    "files = ['/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/03_protein_sequence_properties/polysome_data/Hsu_data_dict.json']\n",
    "path = '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/04_gene_expression'\n",
    "for dirpath, dirnames, filenames in os.walk(path):\n",
    "    for filename in filenames:\n",
    "        if filename.endswith('.json'):\n",
    "            # some of the JSONs we want to keep are found in the probably_junk\n",
    "            # folder, hence I did not exclude the folder from the search. There \n",
    "            # will be extra files though, so will need to filter them out later.\n",
    "            files.append(os.path.join(dirpath, filename))\n",
    "\n",
    "len(files) # 22"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a dataframe with the file paths and feature names  \n",
    "file_paths = list(np.repeat(sorted(files)[:1], 50)) # Hsu 2016 data\n",
    "file_paths = file_paths + list(np.repeat(files[1:], 25)) # all other features\n",
    "file_paths = pd.DataFrame(file_paths).sort_values(0)\n",
    "feat_names = file_paths.iloc[:,0].str.split('/').apply(lambda x: x[-1]).\\\n",
    "    apply(lambda x: x.split('_dict.json')[0].replace('_', ' ')) # get the feature name from the JSON file name\n",
    "feat_names = feat_names.apply(lambda x: x.replace(' dictionary.json', ''))\n",
    "file_paths.insert(1, 'Feature name', feat_names) # insert the feature name into the dataframe\n",
    "file_paths.columns = ['File path', 'Feature name']\n",
    "file_paths['Feature name'] = file_paths['Feature name'].str.lower()\n",
    "\n",
    "# Map the file paths to the master list of features\n",
    "feat_map = {'Abiotic stress differential expression breadth, root':'abiotic root breadth up down',\n",
    "       'Abiotic stress downregulated expression breadth, root':'abiotic-root expr breadth downonly',\n",
    "       'Abiotic stress upregulated expression breadth, root':'abiotic-root expr breadth uponly',\n",
    "       'Abiotic stress differential expression breadth, shoot':'abiotic shoot breadth up down',\n",
    "       'Abiotic stress downregulated expression breadth, shoot':'abiotic-shoot expr breadth downonly',\n",
    "       'Abiotic stress upregulated expression breadth, shoot':'abiotic-shoot expr breadth uponly',\n",
    "       'Biotic stress differential expression breadth':'biotic breadth up down',\n",
    "       'Biotic stress downregulated expression breadth':'biotic expr breadth downonly',\n",
    "       'Biotic stress upregulated expression breadth':'biotic expr breadth uponly',\n",
    "       'Hormone treatment differential expression breadth ':'hormone breadth up down',\n",
    "       'Hormone treatment downregulated expression breadth ':'hormone expr breadth downonly',\n",
    "       'Hormone treatment upregulated expression breadth ':'hormone expr breadth uponly',\n",
    "       'Ribosome occupancy':'hsu data', 'RNA-seq expression':'hsu data'}\n",
    "\n",
    "gene_expr_df['NEW Feature name'] = gene_expr_df['Feature name'].replace(feat_map)\n",
    "gene_expr_df['NEW Feature name'] = gene_expr_df['NEW Feature name'].str.lower()\n",
    "\n",
    "func_anno_paths = file_paths.merge(gene_expr_df, left_on='Feature name',\n",
    "    right_on='NEW Feature name', how='right', suffixes=('', '_y')) # merge to master list\n",
    "func_anno_paths.drop(columns=['Feature name'], inplace=True)\n",
    "func_anno_paths.rename(columns={'Feature name_y':'Feature name'}, inplace=True)\n",
    "columns = func_anno_paths.columns.tolist()\n",
    "columns.insert(2, columns.pop(columns.index('Feature name')))\n",
    "func_anno_paths = func_anno_paths.reindex(columns=columns)\n",
    "\n",
    "# Drop duplicates created by merge\n",
    "func_anno_paths = func_anno_paths.drop_duplicates('ML model feature name').reset_index().drop('index', axis=1)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the checklist file\n",
    "func_anno_paths['Transformation'] = func_anno_paths['Transformation'].fillna('None') # set NaNs in Transformation to \"None\"\n",
    "func_anno_paths.insert(0, 'Created by', '')\n",
    "func_anno_paths.insert(1, 'Date created', '')\n",
    "# func_anno_paths.to_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/04_gene_expression_feature_list.csv', index=False)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 05_epigenetics features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_list = pd.read_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_table_1_corrected_KSA_20240729.csv')\n",
    "\n",
    "# Subset the master list to get the protein properties\n",
    "epi_df = master_list.loc[master_list['Feature category']=='Epigenetic modifications',:]\n",
    "\n",
    "# Get all the JSON file paths\n",
    "files = []\n",
    "path = '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/05_epigenetics'\n",
    "for dirpath, dirnames, filenames in os.walk(path):\n",
    "    if (\"probably_junk\" in dirpath) | (\"test\" in dirpath) | (\"methyl_dicts\" in dirpath):\n",
    "        pass\n",
    "    else:\n",
    "        for filename in filenames:\n",
    "            if filename.endswith('.json'):\n",
    "                files.append(os.path.join(dirpath, filename))\n",
    "len(files) # 31\n",
    "\n",
    "# Create a dataframe with the file paths and feature names \n",
    "file_paths = list(np.repeat(sorted(files)[:9], 4)) + list(np.repeat(sorted(files)[30], 4))\n",
    "file_paths = file_paths + list(np.repeat(sorted(files)[9:30], 25))\n",
    "file_paths = pd.DataFrame(file_paths).sort_values(0)\n",
    "\n",
    "feat_names = file_paths.iloc[:,0].str.split('/').apply(lambda x: x[-1]).\\\n",
    "    apply(lambda x: x.split('_dict.json')[0].replace('_', ' ')) # get the feature name from the JSON file name\n",
    "feat_names = feat_names.apply(lambda x: x.replace(' dictionary.json', ''))\n",
    "\n",
    "file_paths.insert(1, 'Feature name', feat_names) # insert the feature name into the dataframe\n",
    "file_paths.columns = ['File path', 'Feature name']\n",
    "file_paths['Feature name'] = file_paths['Feature name'].str.lower()\n",
    "\n",
    "print(len(file_paths), epi_df.shape) # 93 features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map the file paths to the master list of features\n",
    "epi_df.loc[epi_df['Feature name']=='Gene body methylation', 'NEW Feature name'] = 'bodymethylation.mld'\n",
    "\n",
    "func_anno_paths = file_paths.merge(epi_df, left_on='Feature name', right_on='NEW Feature name',\n",
    "    how='left', suffixes=('', '_y')) # merge to master list\n",
    "func_anno_paths.drop(columns=['Feature name'], inplace=True)\n",
    "\n",
    "# Drop duplicates created by merge\n",
    "func_anno_paths = func_anno_paths.drop_duplicates('ML model feature name').reset_index().drop('index', axis=1)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the checklist file\n",
    "func_anno_paths['Transformation'] = func_anno_paths['Transformation'].fillna('None') # set NaNs in Transformation to \"None\"\n",
    "func_anno_paths.insert(0, 'Created by', '')\n",
    "func_anno_paths.insert(1, 'Date created', '')\n",
    "func_anno_paths.rename(columns={'Feature name_y':'Feature name'}, inplace=True)\n",
    "# func_anno_paths.to_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/05_epigenetics_feature_list.csv', index=False)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 06_network_properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_list = pd.read_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/Supplemental_table_1_corrected_KSA_20240729.csv')\n",
    "\n",
    "# Subset the master list to get the protein properties\n",
    "ntwk_df = master_list.loc[master_list['Feature category']=='Network properties',:]\n",
    "\n",
    "# Get all the JSON file paths\n",
    "files = []\n",
    "path = '/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/21_arabidopsis_redundancy/06_network_properties'\n",
    "for dirpath, dirnames, filenames in os.walk(path):\n",
    "    if (\"probably_junk\" in dirpath) | (\"individual_dicts\" in dirpath):\n",
    "        pass\n",
    "    else:\n",
    "        for filename in filenames:\n",
    "            if filename.endswith('.json'):\n",
    "                files.append(os.path.join(dirpath, filename))\n",
    "\n",
    "len(files)\n",
    "\n",
    "# Create a dataframe with the file paths and feature names \n",
    "file_paths = list(np.repeat(files[0], 175)) + list(np.repeat(files[1], 15))\n",
    "file_paths = file_paths + list(np.repeat(files[3], 15))\n",
    "file_paths = pd.DataFrame(file_paths).sort_values(0)\n",
    "\n",
    "feat_names = file_paths.iloc[:,0].str.split('/').apply(lambda x: x[-1]).\\\n",
    "    apply(lambda x: x.split('_dict_011518.json')[0].replace('_', ' ')) # get the feature name from the JSON file name\n",
    "feat_names = feat_names.apply(lambda x: x.replace(' dicts.json', ''))\n",
    "\n",
    "file_paths.insert(1, 'Feature name', feat_names) # insert the feature name into the dataframe\n",
    "file_paths.columns = ['File path', 'Feature name']\n",
    "file_paths['Feature name'] = file_paths['Feature name'].str.lower()\n",
    "\n",
    "print(len(file_paths), ntwk_df.shape) # 93 features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Map the file paths to the master list of features\n",
    "ntwk_df.loc[ntwk_df['Feature name'].str.contains('Co-expression'), 'NEW Feature name'] = 'all clust'\n",
    "ntwk_df.loc[ntwk_df['Feature name'].str.contains('AraNet'), 'NEW Feature name'] = 'aranet'\n",
    "ntwk_df.loc[ntwk_df['Feature name'].str.contains('Protein-protein'), 'NEW Feature name'] = 'ppi'\n",
    "\n",
    "func_anno_paths = file_paths.merge(ntwk_df, left_on='Feature name', right_on='NEW Feature name',\n",
    "    how='left', suffixes=('', '_y')) # merge to master list\n",
    "func_anno_paths.drop(columns=['Feature name'], inplace=True)\n",
    "\n",
    "# Drop duplicates created by merge\n",
    "func_anno_paths = func_anno_paths.drop_duplicates('ML model feature name').reset_index().drop('index', axis=1)\n",
    "func_anno_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make the checklist file\n",
    "func_anno_paths['Transformation'] = func_anno_paths['Transformation'].fillna('None') # set NaNs in Transformation to \"None\"\n",
    "func_anno_paths.insert(0, 'Created by', '')\n",
    "func_anno_paths.insert(1, 'Date created', '')\n",
    "func_anno_paths.rename(columns={'Feature name_y':'Feature name'}, inplace=True)\n",
    "func_anno_paths.to_csv('/home/seguraab/ara-kinase-prediction/data/2021_cusack_data/06_network_properties_feature_list.csv', index=False)\n",
    "func_anno_paths"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "shap",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
